File: telegram-bot/telemetry.py
import os
import psycopg2
from datetime import datetime

_conn = None

def init_telemetry():
    global _conn
    try:
        db_url = os.environ.get("POSTGRES_URL")
        if db_url:
            _conn = psycopg2.connect(db_url)
            _conn.autocommit = True
            print("Telemetry DB connected")
    except Exception as e:
        print("Telemetry DB connection failed:", e)

def emit_telemetry(bot_id, status, currently_thinking, last_task, task_status):
    global _conn
    if not _conn:
        return
    try:
        with _conn.cursor() as cur:
            query = """
                INSERT INTO bot_telemetry (bot_id, status, currently_thinking, last_task, task_status, last_active_at)
                VALUES (%s, %s, %s, %s, %s, NOW())
                ON CONFLICT (bot_id) DO UPDATE SET
                    status = EXCLUDED.status,
                    currently_thinking = EXCLUDED.currently_thinking,
                    last_task = EXCLUDED.last_task,
                    task_status = EXCLUDED.task_status,
                    last_active_at = NOW();
            """
            cur.execute(query, (bot_id, status, currently_thinking, last_task, task_status))
    except Exception as e:
        print("Failed to emit telemetry:", e)

File: userbot/telemetry.js
const { Client } = require('pg');
require('dotenv').config({ path: '../.env' });

const client = new Client({ connectionString: process.env.POSTGRES_URL });
let connected = false;

async function initTelemetry() {
  try {
    await client.connect();
    await client.query(`
      CREATE TABLE IF NOT EXISTS bot_telemetry (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        bot_id VARCHAR(50) UNIQUE NOT NULL,
        status VARCHAR(20) NOT NULL,
        currently_thinking TEXT,
        last_task TEXT,
        task_status VARCHAR(20),
        last_active_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `);
    connected = true;
    console.log("Telemetry DB connected and table verified");
  } catch (err) {
    console.error("Telemetry DB connection failed:", err);
  }
}

async function emitTelemetry(botId, status, currentlyThinking, lastTask, taskStatus) {
  if (!connected) return;
  try {
    const query = `
      INSERT INTO bot_telemetry (bot_id, status, currently_thinking, last_task, task_status, last_active_at)
      VALUES ($1, $2, $3, $4, $5, NOW())
      ON CONFLICT (bot_id) DO UPDATE SET
        status = EXCLUDED.status,
        currently_thinking = EXCLUDED.currently_thinking,
        last_task = EXCLUDED.last_task,
        task_status = EXCLUDED.task_status,
        last_active_at = NOW();
    `;
    await client.query(query, [botId, status, currentlyThinking, lastTask, taskStatus]);
  } catch (err) {
    console.error("Failed to emit telemetry:", err);
  }
}

module.exports = { initTelemetry, emitTelemetry };
